### PR TITLE
feat: add events to contribution lifecycle

### PR DESCRIPTION
**Pull request by tdelabro**
feat: add migration script to emit events for already stored contributions

### CODE REVIEW

**Comment by tdelabro** [comment]
We have a problem:
`new_contribution` can be used both for creating and updating a contribution.
It seems to be the expected behavior, as stated in this test:
```cairo
    let (local contribution1) = contributions.new_contribution(123, 'MyProject', 0, 'validator')
    let (contribution1) = contributions.new_contribution(123, 'MyProject', 0, 'validator')  # adding twice the same to test update
```
But right now, we are emitting the same event, `ContributionCreated`, in both cases.
We have to either:
- prohibit this behavior in `new_contribution` and introduce a `update_contribution` method
- keep this behavior but introduce a storage check in order to distinguish between `create` and `update` calls, and emit the correct event. 

What would you prefer?

File: contracts/onlydust/deathnote/core/contributions/library.cairo [hunk 1]
```diff
@@ -39,6 +39,29 @@ struct Contribution:
...
+@event
+func ContributionAssigned(contribution_id: felt, contributor_id: Uint256):
+end
+
+@event
+func ContributionUnassigned(contribution_id: felt):
+end
+
+@event
+func ContributionValidated(contribution_id: felt):
```
Comment (Bernardstanislas) [inline contracts/onlydust/deathnote/core/contributions/library.cairo:58]: I can't remember if we chose validated or completed as the last status?

**Review by Bernardstanislas — APPROVED** [review]

**Comment by Bernardstanislas** [comment]
I would prefer being explicit over implicit, preventing this behavior. It is not visible in the name of the function and the event that they can be used as updates. So preventing is better imho

**Comment by Bernardstanislas** [comment]
What is your recommendation @tdelabro ?

File: contracts/onlydust/deathnote/core/contributions/library.cairo (hunk 1 above)
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/contributions/library.cairo:58]: The last status is `COMPLETED`, but teh action to go into this status is `validate()`
Are events supposed to be named after statuses or actions ?

**Review by AnthonyBuisset — APPROVED** [review]

**Comment by tdelabro** [comment]
My opinion is that it's a misleading pattern. It already started to cause trouble.
We should create two separate functions, at least at the `@view` level. That will check if either the contribution already exists or it doesn't, and then internally call the same `unchecked_write_contribution`.

**Comment by tdelabro** [comment]
Events are based on actions.
The same status could be the result of multiple different actions.

**Review by AnthonyBuisset — APPROVED** [review]

**Pull request merged**